### PR TITLE
Kwxm/cek costing profiling (SCP-612)

### PR DESCRIPTION
**Pull request by kwxm**
This reports some results on profiling some Plutus Core programs, especially how budgeting affects execution time.  I think we can say that it covers SCP-612.  It consists of a single markdown file together with some graphs.  The main document is in https://github.com/input-output-hk/plutus/blob/kwxm/cek-costing-profiling/notes/plutus-core/cek-budgeting-profiling/CekProfiling.md

Update: see also https://github.com/input-output-hk/plutus/blob/kwxm/cek-costing-profiling/notes/plutus-core/cek-budgeting-profiling/CostingDetails.md .

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `$(nix-build default.nix -A pkgsLocal.updateMaterialized)` to update the materialized Nix files
   - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [ ] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [ ] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Review by michaelpj — APPROVED** [review]
Thanks, very interesting. I'd still like to try something like using `STRef`s for all the budgeting stuff.

File: notes/plutus-core/cek-budgeting-profiling/CostingDetails.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+The results for the validation benchmarks above show that apparently small
+changes in the costing code can make quite a large difference to the execution
+times.  In column C all of our sample validations are comfortably under the 20ms
+budget that's been mentioned in the past (at least on my machine), and some of
+the nofib examples are approaching a 3x speedup.
+
+Here are some suggestions about how to speed things up even more:
+
+* Remove the lens operations and make everything explicit: this may make it
+  easier to see what's going on and help avoid memory leaks.
```
Comment (effectfully) [inline notes/plutus-core/cek-budgeting-profiling/CostingDetails.md:224]: This is a great idea, I believe.

File: notes/plutus-core/cek-budgeting-profiling/CostingDetails.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+            Counting -> modifying exBudgetStateTally (<> (ExTally (singleton key budget)))
+            Restricting resb -> 
+                when (exceedsBudget resb newBudget) $
+                    throwingWithCause _EvaluationError
+                        (UserEvaluationError $ CekOutOfExError resb newBudget)
+                        Nothing
+```
+
+eliminates the leak again, but it's not clear to me why.  Possibly `modifying`
+is accumulating costs lazily and updating `_exBudgetStateBudget` prevents this
```
Comment (effectfully) [inline notes/plutus-core/cek-budgeting-profiling/CostingDetails.md:115]: As per a slack discussion, yeah, `modifying` calls the lazy `modify` instead of the strict `modify'`.

File: notes/plutus-core/cek-budgeting-profiling/CostingDetails.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+```
+
+This does reduce execution times noticeably in `Restricting` mode (see table
+later), but for some reason it reintroduces a memory leak in `Counting` mode:
+many of the `nofib` examples fail to complete because they consume large amounts
+of memory very quickly.  Changing the code to 
+
+```
+    spendBudget key budget = do
+        newBudget <- exBudgetStateBudget <%= (<> budget)
```
Comment (effectfully) [inline notes/plutus-core/cek-budgeting-profiling/CostingDetails.md:103]: This does not look like it's strict enough, however I guess your guess is probably correct: we disturb the record by only accessing `_exBudgetStateBudget` at which point all previous thunks get forced and then we create a new one by modifying that field with `(<> budget)`. And that last thunk will be forced at the next iteration.

File: notes/plutus-core/cek-budgeting-profiling/CostingDetails.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+Here are some suggestions about how to speed things up even more:
+
+* Remove the lens operations and make everything explicit: this may make it
+  easier to see what's going on and help avoid memory leaks.
+* Try to make `Restricting` mode and `Counting` mode completely separate
+  so that we don't have to examine the mode every time we want to spend
+  some of the budget.  Using a typeclass makes this a bit difficult because
+  we can only have one instance per type.  Could we (a) provide two methods,
+  one for each mode, or (b) dispense with the `SpendBudget` class altogether
+  and pass an accounting function as a parameter to the machine?  The latter would
```
Comment (effectfully) [inline notes/plutus-core/cek-budgeting-profiling/CostingDetails.md:230]: I like the "pass an accounting function" strategy.

**Review by effectfully — APPROVED** [review]
:+1:

File: notes/plutus-core/cek-budgeting-profiling/CostingDetails.md (hunk 2 above)
Comment (kwxm) [reply notes/plutus-core/cek-budgeting-profiling/CostingDetails.md:115]: :+1: I've added a note about that.

File: notes/plutus-core/cek-budgeting-profiling/CostingDetails.md (hunk 3 above)
Comment (effectfully) [reply notes/plutus-core/cek-budgeting-profiling/CostingDetails.md:103]: Yeah, I believe this is exactly what happens. If we update the code to the following:

```haskell
counterVar :: IORef Int
counterVar = unsafePerformIO $ newIORef 0
{-# NOINLINE counterVar #-}

instance (Eq fun, Hashable fun, ToExMemory term) =>
            SpendBudget (CekCarryingM term uni fun) fun (ExBudgetCategory fun) term where
    spendBudget key budget = do
        let !counter = unsafePerformIO $ readIORef counterVar
        let !_ = unsafePerformIO $ modifyIORef' counterVar succ
        newBudget <- trace (show counter ++ ":1") exBudgetStateBudget <%= (<> trace (show counter ++ ":2") budget)
        mode <- asks cekEnvBudgetMode
        case mode of
            Counting -> modifying (trace (show counter ++ ":3") exBudgetStateTally) (<> trace (show counter ++ ":4") (ExTally (singleton key budget)))
            Restricting -> <...>
```

then this gives us the following output: 

```
0:1
1:1
0:3
0:2
0:4
2:1
1:3
1:2
1:4
3:1
2:3
2:2
2:4
<...>
```

where a number before `:` is an iteration counter and a number after `:` indicates a particular place in the code. The beginning of this sequence is

```
0:1
1:1
```

which demonstrates that `<%=` forces its first argument, but does not actually update the record (that becomes a thunk). `modifying` does not even force its first argument, let alone update the record. And so at the zeroth iteration we've only forced the record by accessing it in `<%=` and constructed a bunch of thunks. At the first iteration we access the record again and this causes all the thunks created previously to be forced as well, giving us

```
0:3
0:2
0:4
```

And then the process continues: accessing the record via `<%=` forces thunks created at the previous iteration.

So basically `<%=` and `modifying` both modify the state, but have different strictness properties.

File: notes/plutus-core/cek-budgeting-profiling/CostingDetails.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+  avoid overhead due to calling typeclass methods (see [CekProfiling.md](./CekProfiling.md)),
+  so might be better.
+* In `Restricting` mode, decrement the cost at every step and fail it if goes
+  below zero.  It's conceivable that it'd be cheaper to check whether an Integer
+  is negative rather than compare two (possibly large) Integers.
+* Maybe don't use a monoid to accumulate costs in `Restricting` mode.  Michael
```
Comment (michaelpj) [inline notes/plutus-core/cek-budgeting-profiling/CostingDetails.md:247]: The thing that actually worries me is the constant mutation of the map. Mutating a `Map X Integer` to increment an integer requires copying ~1/2 of the internal structure depending on where the key is. The nice thing about a `Map X (STRef Integer)` is that the *map* never chagnes, only the mutable cells *in* the map change.

**Review by michaelpj — APPROVED** [review]

File: notes/plutus-core/cek-budgeting-profiling/CekProfiling.md [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+generally faster, taking about 75-95% of the time required to deserialise CBOR.
+
+Quite a lot of time seems to be spent looking up names in environments and
+inserting new names (about 80% for crowdfunding, 60% for zerocoupon, 25% for
+primetest).  Using de Bruijn levels could help with this: we already know that
+de Bruijn _indices_ help to reduce script sizes, and presumably the same would
+apply for levels.  However, if we can compute directly with de Bruijn levels it
+might be possible to look things up more rapidly: since we know in advance how
+deep recursion will be I think we could pre-allocate an array of mutable
+variables to store the terms referred to by the levels, and this would allow
```
[the remark this answers is not in the document]
Comment (kwxm) [reply notes/plutus-core/cek-budgeting-profiling/CekProfiling.md]: Yeah, I'm not sure about that. I thought we could have a sort of stack of bindings, but maybe that'd require us to start re-indexing variables inside terms, which we want to avoid.  We should probably try to implement something, then any problems would become more apparent.

[the remark this answers is not in the document]
Comment (michaelpj) [reply notes/plutus-core/cek-budgeting-profiling/CekProfiling.md]: Sure, but then you either have
- A stack of mutable arrays (which you then can't pre-allocate); or
- A mutable array of stacks (in which case you probably aren't benefiting that much from the mutability; or
- You use machine frames to restore old variable bindings when you leave (in which case you're using the GHC stack and you're anyway maybe making something more like a stack machine)

[the remark this answers is not in the document]
Comment (kwxm) [reply notes/plutus-core/cek-budgeting-profiling/CekProfiling.md]: That's not exactly what I meant.  I had a vague idea that you could have one big long mutable array and effectively use it as a stack, advancing further along every time you add a new binding and retreating when that binding goes out of scope (perhaps exposing an earlier binding for the same variable).  I didn't think it through though, and I think I'd have to start implementing something before I could see if that made sense.

File: notes/plutus-core/cek-budgeting-profiling/CostingDetails.md (hunk 5 above)
Comment (kwxm) [reply notes/plutus-core/cek-budgeting-profiling/CostingDetails.md:259]: In `Restricting` mode (which is what we want on the chain), we only have a record with two entries (CPU limit and memory limit); I agree that just passing two mutable integers about might well help though.  The map's only used in `Counting` mode, and the efficiency of that isn't critical (but let's make it efficient if we can).
